File: src/main/java/com/learn/todolist/infra/TokenProvider.java
package com.learn.todolist.infra;

import com.learn.todolist.utils.EnvUtils;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.stereotype.Component;

import java.util.Date;
import java.util.Map;
import java.util.UUID;

@RequiredArgsConstructor
@Component
public class TokenProvider {

    private final String secret = EnvUtils.getEnv("jwt-secret-key");
    private final long accessExpirationHours = Long.parseLong(EnvUtils.getEnv("jwt-access-expiration-hours"));
    private final String issuer = EnvUtils.getEnv("jwt-issuer");

    private final byte[] key = Keys.hmacShaKeyFor(secret.getBytes()).getEncoded();

    public String generateToken(UUID userId) {
        Map<String, Object> claims = Map.of("type", "access");
        return Jwts.builder()
                .setClaims(claims)
                .setSubject(String.valueOf(userId))
                .setIssuer(issuer)
                .setIssuedAt(new Date())
                .setExpiration(new Date(System.currentTimeMillis() + accessExpirationHours * 60 * 60 * 1000))
                .signWith(Keys.hmacShaKeyFor(key))
                .compact();
    }

    public boolean validateToken(String token) {
        try {
            Jwts.parserBuilder()
                    .setSigningKey(Keys.hmacShaKeyFor(key))
                    .build()
                    .parseClaimsJws(token);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public UsernamePasswordAuthenticationToken authenticateToken(String token) {
        var claims = Jwts.parserBuilder()
                .setSigningKey(Keys.hmacShaKeyFor(key))
                .build()
                .parseClaimsJws(token)
                .getBody();
        return new UsernamePasswordAuthenticationToken(claims.getSubject(), null, null);
    }
}

File: src/main/kotlin/com/todolist/infra/security/TokenProvider.kt
package com.todolist.infra.security

import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.Date
import javax.crypto.spec.SecretKeySpec
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.PropertySource
import org.springframework.stereotype.Component
import com.todolist.domain.user.entity.User
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import java.sql.Timestamp
import java.time.LocalDateTime

@PropertySource("classpath:jwt.yml")
@Component
class TokenProvider(
    @Value("\${secret-key}")
    private val secretKey: String,
    @Value("\${access-expiration-hours}")
    private val accessExpirationHours: Long,
    @Value("\${issuer}")
    private val issuer: String
) {
    fun createToken(user: User): String {
        val accessClaim: Claims = Jwts.claims().apply {
            put("type", "access")
        }
        val accessToken: String = Jwts.builder()
            .signWith(SecretKeySpec(secretKey.toByteArray(), SignatureAlgorithm.HS256.jcaName))
            .setClaims(accessClaim)
            .setIssuer(issuer)
            .setSubject(user.id.toString())
            .setIssuedAt(Timestamp.valueOf(LocalDateTime.now()))
            .setExpiration(Date.from(Instant.now().plus(accessExpirationHours, ChronoUnit.HOURS)))
            .compact()
        return accessToken
    }

    fun validateToken(token: String): Boolean {
        return try{
            val claims = Jwts.parserBuilder()
                .setSigningKey(SecretKeySpec(secretKey.toByteArray(), SignatureAlgorithm.HS256.jcaName))
                .build()
                .parseClaimsJws(token)
            claims.body.expiration.after(Date.from(Instant.now()))
        } catch(e: Exception) {
            return false
        }
    }

    fun authenticateToken(token: String): Authentication {
        val claims = Jwts.parserBuilder()
            .setSigningKey(SecretKeySpec(secretKey.toByteArray(), SignatureAlgorithm.HS256.jcaName))
            .build()
            .parseClaimsJws(token)
        val authorities = listOf(SimpleGrantedAuthority("ROLE_USER"))
        return UsernamePasswordAuthenticationToken(claims.body.subject, null, authorities)
    }
}